### PR TITLE
feat: streamable procedure support

### LINKED ISSUE #4477: feat: Streaming Mutations / Queries
### Describe the feature you'd like to request

Many upstream APIs (e.g. OpenAI's GPT) stream responses to the client. It would be great if TRPC supported such streaming requests.

### Describe the solution you'd like to see

I'd imagine this would look like a mutation or query where the resolver is an async generator. Example:

```typescript
publicProcedure
   .streamingMutation(async function* ({ input, ctx }) {
     for(let x=0x;<10;x++) {
       yield x;
       await sleep(100);
     }
   });
```

### Describe alternate solutions

An alternative is to set up subscriptions and map the mutation inputs to some identifier in the subscription. The problem with this approach is (a) it depends on websockets, which are difficult to support from an infrastructure perspective, (b) the race conditions between establishing a socket connection and corresponding this with mutations/queries can be miserable.

### Additional information

I'd mostly like to understand any suggestions for how to approach this and whether you'd support a PR. Creating a `streamingMutation` procedure type, for instance, is fairly invasive but may be the correct approach.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4477">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4477/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by helmturner**
revert(server): roll back node adapter to 0aaf147;
revert(server): roll back fetch adapter to 0aaf147;
revert(server): roll back fastify adapter to ababea8;

Closes #4477

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

- Reverts adapters to pre-httpBatchLink versions and extracts response-body creation logic into separate modules
- Enables ReadableStream bodies to be created by and returned from `resolveHTTPResponse`
- Sets the stage for eagerly streaming queries 

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Comment by helmturner** [comment]
On further, further investigation... I'm not actually reading out of order, at least in most cases. Everything works pretty well, actually. There are plenty of successful tests for both batch and single procedure calls. Save for some edge cases and some adapter issues, looks like this is working? Merging main brought it to 50 failing tests, but I made some small changes and cut that to 36 failing tests - a personal best, so far!

File: packages/server/src/http/buildResponse.ts [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        }
+
+        const chunks = getRemainingChunks(envelope.result.data);
+        //XXX
+        if (style !== 'event-stream' && style !== 'json-stream') {
+          // We're not streaming, so we need to resolve the generator and return the complete result
+          let done;
+          let value;
+
+          while (!done) ({ done, value } = await chunks.next());
```
Comment (KATT) [inline packages/server/src/http/buildResponse.ts:197]: I think we should just throw here instead and have the rule if you request an async generator, you have to use a transport/http link that supports it

Comment (helmturner) [reply packages/server/src/http/buildResponse.ts:197]: It seems intuitive that returning a generator from a tRPC resolver wouldn't work without additional configuration. I guess my thought was preserving compatibility in the case you have multiple clients in heterogenous runtimes with varying degrees of streaming support, e.g. mobile and web.

My toy use-case for the SSE side of things was the GPT4 API, but blocking on a mobile device until the response completes is a pretty crap experience in that case, so it's probably an unrealistic edge-case unless there's some other scenario where blocking for an otherwise-streamed response is acceptable UX.

**Comment by KATT** [comment]
#4911
